File: printer_server/drivers/wintech/wintech_snip.js
var enable_wintech_start_stop_buttons = function () {
    $('#wintech-start-btn').prop('disabled', false);
    $('#wintech-stop-btn').prop('disabled', false);
}

var disable_wintech_start_stop_buttons = function () {
    $('#wintech-start-btn').prop('disabled', true);
    $('#wintech-stop-btn').prop('disabled', true);
}

$(document).ready(function () {
    // Handles to user inputs
    var WintechLedPowerSliderElement = document.getElementById("wintech-led-power-slider");
    var WintechLedPowerLabelElement = document.getElementById("wintech-led-power-label");
    var WintechExposureElement = document.getElementById("wintech-exposure-txt");

    // Set initial LED power slider label value
    WintechLedPowerLabelElement.innerHTML = WintechLedPowerSliderElement.value; // Display the default slider value

    // Update the WintechLedPowerLabelElement with the current slider value
    WintechLedPowerSliderElement.oninput = function () {
        WintechLedPowerLabelElement.innerHTML = this.value;
    }

    // Light engine control stop button click function
    $("#wintech-stop-btn").click(function () {
        socket.emit("wintech_stop");
    });

    // Exposure text input change function
    $('#wintech-exposure-txt').on('change', function () {
        exposure = WintechExposureElement.value;

        // Validate user input. Only allows positive integers > 0
        if (/^\d+$/.test(exposure) && exposure > 0) {
            WintechExposureElement.classList.remove("is-invalid")
            enable_wintech_start_stop_buttons();
        } else {
            WintechExposureElement.classList.add("is-invalid")
            disable_wintech_start_stop_buttons();
        }
    })

    $("#wintech-repeat-chkbx").click(function () {
        var repeatCheckboxElement = document.getElementById("wintech-repeat-chkbx");
        var repeat = Number(!repeatCheckboxElement.checked);
        if (repeat == 0) {
            WintechExposureElement.classList.remove("is-invalid")
            $('#wintech-exposure-txt').prop('disabled', true);
        } else {
            $('#wintech-exposure-txt').prop('disabled', false);
        }
    });

    // Light engine control start button click function
    $("#wintech-start-btn").click(function () {
        var repeatCheckboxElement = document.getElementById("wintech-repeat-chkbx");
        var exposure = WintechExposureElement.value;
        var repeat = Number(!repeatCheckboxElement.checked);
        var ledPower = WintechLedPowerSliderElement.value;

        if (!/^\d+$/.test(exposure) && !exposure > 0) {
            if (repeat == 1) {
                WintechExposureElement.classList.add("is-invalid")
                return
            } else {
                exposure = 1
            }
        }
        socket.emit("wintech_start", { "repeat": repeat, "exposure": exposure, "ledPower": ledPower });
    });

    socket.on("update_wintech_led_status", function (message) {
        var statusElement = document.getElementById("wintech-status");
        if (message == true) {
            statusElement.classList.remove("invisible")
        } else {
            statusElement.classList.add("invisible")
        }
        $('#wintech-status').prop('disabled', message);
    });

});

File: printer_server/drivers/wintech/wintech_snip.py
from printer_server.extensions import socketio
from printer_server.hardware_configuration.hardware_configuration import driver_handles

wintech = driver_handles.wintech


@socketio.on("wintech_stop", namespace="/manual")
def wintechStop():
    """Turn off the LED in the light engine."""
    wintech.stop_sequencer()
    socketio.emit("wintech_update_led_state", False, namespace="/manual")
    socketio.emit("wintech_done", namespace="/manual")


@socketio.on("wintech_start", namespace="/manual")
def wintechProject(message):
    """Project the image with the given settings."""
    ledPower = int(message["ledPower"])
    repeat = int(message["repeat"])
    exposure = int(message["exposure"])
    socketio.emit("wintech_update_led_state", True, namespace="/manual")
    wintech.project(exposure, ledPower, repeat)
    if repeat != 0:
        socketio.emit("wintech_update_led_state", False, namespace="/manual")
    socketio.emit("wintech_done", namespace="/manual")


# @socketio.on("wintech_get_status", namespace="/manual")
# def wintechStatus():
#     socketio.emit(
#         "wintech_return_status",
#         wintech.read_all_status(),
#         namespace="/manual"
#     )


def getLedStatus():
    return wintech.led_on
